### PR TITLE
add nb instances and timer in mail

### PR DESCRIPTION
**Pull request by Hoshiyo**

### CODE REVIEW

File: fabfile/utils.py [hunk 1]
```diff
@@ -373,3 +379,33 @@ def show_dead_kraken_status(warn_dict, show=False):
...
+    def register_start(self, service):
+        self.time_dict[service] = datetime.datetime.now()
+
+    def register_end(self, service):
+        self.time_dict[service] = datetime.datetime.now() - self.time_dict[service]
+
+    def get_time_diff(self, service):
+        return self.time_dict[service]
+
+    def show_time_deploy(self, show=False):
```
Comment (antoine-de) [inline fabfile/utils.py]: In term of responsability I don't think this needs to be in TimeDiff.

TimeDiff just needs time some stuff regardless of `bina`, `kraken` or else.
I think this code should maybe a free fonction called at the end of the bina ?

File: fabfile/utils.py [hunk 2]
```diff
@@ -357,6 +358,11 @@ def start_or_stop_with_delay(service, delay, wait, start=True, only_once=False,
 
 def show_dead_kraken_status(warn_dict, show=False):
     status = ''
+    count = 0
+    while count < len(env.instances.values()):
```
Comment (antoine-de) [inline fabfile/utils.py]: hum I think you can just return len(env.instances.values()) no ?

File: fabfile/utils.py [hunk 3]
```diff
@@ -357,6 +358,11 @@ def start_or_stop_with_delay(service, delay, wait, start=True, only_once=False,
...
+        count += 1
+    if count > 0:
+        status += "\nThere are {} instances.".format(count)
```
Comment (antoine-de) [inline fabfile/utils.py]: EDIT: you should just do
status = ""\nThere are {} instances.".format(len(env.instances)) if env.instance else ""

(or put the message there are 0 instances, I think it's usefull)

File: fabfile/utils.py (hunk 2 above)
Comment (xlqian) [inline fabfile/utils.py]: `len(env.instances)` should be enough, no?

Comment (xlqian) [inline fabfile/utils.py]: and count=0 and `count < len(env.instances.values())` ? if `len(instances.values()) == 0` we'll never enter in the loop, is that wanted?

File: fabfile/tasks.py [hunk 4]
```diff
@@ -133,16 +138,21 @@ def upgrade_all(up_tyr=True, up_confs=True, kraken_wait=True, check_version=True
...
     else:
         execute(kraken.swap_all_data_nav)
+        time_dict.register_start('kraken')
         execute(upgrade_kraken, kraken_wait=kraken_wait, up_confs=up_confs)
+        time_dict.register_end('kraken')
         execute(upgrade_jormungandr, up_confs=up_confs)
+    time_dict.register_end('total_deploy')
     warn_dict = jormungandr.check_kraken_jormun_after_deploy()
     status = show_dead_kraken_status(warn_dict, show=True)
+    status += show_time_deploy(time_dict)
```
Comment (xlqian) [inline fabfile/tasks.py]: Line 154 and Line 155 can be merged

Comment (xlqian) [inline fabfile/tasks.py]: `status` is only called if `send_mail` is true, you can move `status` in the block of `if send_mail`

Comment (francois-vincent) [inline fabfile/tasks.py]: no because it has side-effet = print !!

**Pull request merged**